File: yasir-truffle/src/main/java/com/github/overmind/yasir/Simple.java
package com.github.overmind.yasir;

import com.github.overmind.yasir.ast.*;
import com.github.overmind.yasir.value.BareFunction;
import com.oracle.truffle.api.frame.*;

public class Simple {
    @SafeVarargs
    public static <A> A[] array(A... args) {
        return args;
    }

    public static Expr makeBench(int count, long warmupN, long n, BareFunction c) {
        FrameDescriptor mainFd = new FrameDescriptor();
        FrameSlot funcSlot = mainFd.addFrameSlot("funcSlot");
        FrameSlot nSlot = mainFd.addFrameSlot("n");
        Expr main = MkLambda.create("main", array(nSlot), array(funcSlot),
                Begin.create(
                        Vars.writeBox(funcSlot, PrimOp.lit(c)),

                        // Warmup. Loops needs 2 more times for OSR and the function body.
                        ApplyNode.unknown(Vars.readBox(funcSlot), PrimOp.lit(warmupN)),
                        ApplyNode.unknown(Vars.readBox(funcSlot), PrimOp.lit(warmupN)),
                        ApplyNode.unknown(Vars.readBox(funcSlot), PrimOp.lit(warmupN)),

                        // Actual bench.
                        PrimOp.bench(
                            ApplyNode.unknown(Vars.readBox(funcSlot), Vars.read(nSlot)),
                            count
                        )
                ),
                mainFd);

        return ApplyNode.unknown(main, PrimOp.lit(n));
    }

}

File: yasir-truffle/src/main/scala/Simple.scala
import com.oracle.truffle.api.frame.FrameDescriptor

object Example {
  def makeFibo(n: Int): Expr = {
    val mainFd = Rt.createEnvDescr()
    val fiboFd = Rt.createEnvDescr()
    val fiboSlot = mainFd.addFrameSlot("fiboSlot")
    val nSlot = fiboFd.addFrameSlot("n")

    val fiboBody = MkLambda("fibo", Array(nSlot),
      If(
        LessThan(ReadVar(nSlot), ConstInt(2)),
        ReadVar(nSlot),
        Add(
          Apply(ReadVar(fiboSlot, 1), Array(Sub(ReadVar(nSlot), ConstInt(1)))),
          Apply(ReadVar(fiboSlot, 1), Array(Sub(ReadVar(nSlot), ConstInt(2)))))),
      fiboFd)

    val main = MkLambda("main", Array(),
      Begin(
        Array(
          WriteVar(fiboBody, fiboSlot),
          Apply(ReadVar(fiboSlot), Array(ConstInt(n))))),
      mainFd)

    Apply(main, Array())
  }
}
